### DISCUSSION TITLE
I accidentally removed 3 weeks worth of commits, need help.

### DISCUSSION THREAD

**Discussion by MaxKrekeQuilar** [community/community#29149] [Other Feature Feedback, Questions, & Ideas]
Hello, i might have done a colossal fuck up:
I have been away from work for 3 weeks and when i came back today, i accidentally pushed my old outdated version of the software to github.
When i noticed, i tried looking up how to remove the most recent commit and found this:
https://stackoverflow.com/questions/8225125/remove-last-commit-from-remote-git-repository
When i tried that, it didnt only remove the most recent commit, but all the commits that had been made since i had started my break from work. Is there any way to undo this?

**Comment by airtower-luna** [comment]
Possibly. It'd be easiest if you still have a branch or tag (possibly remote) with the right commits, but I guess in that case you wouldn't be asking. :sweat_smile: 

Both local repositories and GitHub don't immediately delete commits that aren't referenced any more, e.g. because a branch was overwritten by a force-push. The trick is to find the right commit ID, and then you can access the commit using its ID to start a new branch from there, or reset one of your branches to that commit. I'd start with the "new branch" option to avoid additional mistakes, you can always reset another branch later if you're happy with the result.

For your local repository you can use [`git reflog`](https://git-scm.com/docs/git-reflog) to see what commit has been your current HEAD and on what branch (if any) a pretty long time into the past. If that doesn't help you can search for "dangling" commits using [`git fsck`](https://git-scm.com/docs/git-fsck). [Git Internals - Maintenance and Data Recovery](https://git-scm.com/book/en/v2/Git-Internals-Maintenance-and-Data-Recovery) describes the process in detail.

For the GitHub repository it might be more tricky, if you don't happen to have the commit ID in your history or something. Worst case you can try to [contact support](https://support.github.com/request) and ask if they can restore from a backup.

Any sufficiently recent copy of the repository might be a source of the right commit ID, if you know of any other than your local one. E.g. on another computer you have, a backup disk, a colleague's computer, a fork on GitHub…
